language/stdlib/modules/libra_system.move

```diff
@@ -186,69 +196,186 @@ module LibraSystem {
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        let account_address = Transaction::sender();
+
+        // Ensure that this address is already a validator
         Transaction::assert(
-            !Vector::contains(
-                &validator_set_ref.removals,
-                &account_address,
-            ),
-            22
+            is_validator_(&account_address, &validator_set_ref.validators),
+            21
         );
 
-        // Add to removals
-        Vector::push_back(
-            &mut validator_set_ref.removals,
-            account_address,
+        ValidatorConfig::rotate_consensus_pubkey(consensus_pubkey);
+
+        let validator_index = get_validator_index(
+            &validator_set_ref.validators,
+            account_address
+        );
+
+        if(copy_validator_info(Vector::borrow_mut(&mut validator_set_ref.validators, validator_index))) {
+            reconfigure_();
+        }
+    }
+
+    public fun rotate_validator_network_identity_pubkey(
+        validator_network_identity_pubkey: vector<u8>
+    ) acquires DiscoverySet {
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        let account_address = Transaction::sender();
+
+        ValidatorConfig::rotate_validator_network_identity_pubkey(validator_network_identity_pubkey);
+
+        let validator_index = get_discovery_index(
+            &discovery_set_ref.discovery_set,
+            account_address
+        );
+
+        if(copy_discovery_info(Vector::borrow_mut(&mut discovery_set_ref.discovery_set, validator_index))) {
+            emit_discovery_set_change();
+        }
+    }
+
+    public fun rotate_validator_network_address(
+        validator_network_address: vector<u8>
+    ) acquires DiscoverySet {
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        let account_address = Transaction::sender();
+
+        ValidatorConfig::rotate_validator_network_address(validator_network_address);
+
+        let validator_index = get_discovery_index(
+            &discovery_set_ref.discovery_set,
+            account_address
+        );
+
+        if(copy_discovery_info(Vector::borrow_mut(&mut discovery_set_ref.discovery_set, validator_index))) {
+            emit_discovery_set_change();
+        }
+    }
+
+    fun reconfigure_() acquires ValidatorSet {
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        let current_block_time = LibraTimestamp::now_microseconds();
+
+        validator_set_ref.last_reconfiguration_time = current_block_time;
+
+        // Only one reconfiguration allowed per block for now. We might want to increase the threshold here.
+        Transaction::assert(current_block_time > validator_set_ref.last_reconfiguration_time, 23);
+        emit_reconfiguration_event();
+    }
+
+    fun emit_reconfiguration_event() acquires ValidatorSet {
```

Comment:
Comment explaining that we have organized it this way so we can call it from genesis?